PR: Activate conda environment prior to kernel start

goanpeca opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes


* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)


<!--- Explain what you've done and why --->




### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #9077
Fixes spyder-ide/spyder-kernels#89


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: @goanpeca 

<!--- Thanks for your help making Spyder better for everyone! --->

goanpeca commented:
@CAM-Gerlach could you check if this PR fixes  #9077 
This should now properly activate conda envs prior to kernel execution

CAM-Gerlach commented:
I have a flight early tomorrow morning and I'm at a confrence but if I don't do it tonight I'll be next availible tomorrow afternoon/evening.

CAM-Gerlach commented:
Testing in work...

goanpeca commented:
Ok could you share conda list? Of the encironment and the spyder environment?

Thanks!

CAM-Gerlach commented:
@goanpeca As mentioned, it happens on every Spyder environment and working environment that I test, including regardless of the DLL search path modification. I can't confirm on a freshly installed Spyder 4.0.1 env because my internet connection is abysmally slow here at the conference, here's a fresh rc3 env updated to 4.0.1:

<details>

```
# packages in environment at C:\Miniconda3\envs\spyder-test:
#
# Name                    Version                   Build  Channel
alabaster                 0.7.12                     py_0
argh                      0.26.2                py37_1001    spyder-ide
asn1crypto                1.2.0                    py37_0
astroid                   2.3.3                    py37_0
atomicwrites              1.3.0                    py37_1
attrs                     19.3.0                     py_0
autopep8                  1.4.4                      py_0
babel                     2.7.0                      py_0
backcall                  0.1.0                    py37_0
bcrypt                    3.1.7            py37he774522_0
bleach                    3.1.0                      py_0
ca-certificates           2019.11.27                    0
certifi                   2019.11.28               py37_0
cffi                      1.13.2           py37h7a1dbc1_0
chardet                   3.0.4                 py37_1003
cloudpickle               1.2.2                      py_0
colorama                  0.4.1                      py_0
cryptography              2.8              py37h7a1dbc1_0
decorator                 4.4.1                      py_0
defusedxml                0.6.0                      py_0
diff-match-patch          20181111                   py_0    spyder-ide
docutils                  0.15.2                   py37_0
entrypoints               0.3                      py37_0
flake8                    3.7.9                    py37_0
future                    0.18.2                   py37_0
icu                       58.2                 ha66f8fd_1
idna                      2.8                      py37_0
imagesize                 1.1.0                      py_0
importlib_metadata        0.23                     py37_0
intervaltree              3.0.2                      py_0    spyder-ide
ipykernel                 5.1.3            py37h39e3cac_0
ipython                   7.9.0            py37h39e3cac_0
ipython_genutils          0.2.0                    py37_0
isort                     4.3.21                   py37_0
jedi                      0.14.1                   py37_0    spyder-ide
jinja2                    2.10.3                     py_0
jpeg                      9b                   hb83a4c4_2
jsonschema                3.2.0                    py37_0
jupyter_client            5.3.4                    py37_0
jupyter_core              4.6.1                    py37_0
keyring                   19.2.0                   py37_0
lazy-object-proxy         1.4.3            py37he774522_0
libpng                    1.6.37               h2a8f88b_0
libsodium                 1.0.16               h9d3ae62_0
libspatialindex           1.9.3                h33f27b4_0
markupsafe                1.1.1            py37he774522_0
mccabe                    0.6.1                    py37_1
mistune                   0.8.4            py37he774522_0
more-itertools            7.2.0                    py37_0
nbconvert                 5.6.1                    py37_0
nbformat                  4.4.0                    py37_0
numpydoc                  0.9.1                      py_0
openssl                   1.1.1d               he774522_3
packaging                 19.2                       py_0
pandoc                    2.2.3.2                       0
pandocfilters             1.4.2                    py37_1
paramiko                  2.6.0                    py37_0
parso                     0.5.1                      py_0
pathtools                 0.1.2                      py_1    spyder-ide
pexpect                   4.7.0                    py37_0
pickleshare               0.7.5                    py37_0
pip                       19.3.1                   py37_0
pluggy                    0.13.1                   py37_0
prompt_toolkit            2.0.10                     py_0
psutil                    5.6.5            py37he774522_0
pycodestyle               2.5.0                    py37_0
pycparser                 2.19                       py_0
pydocstyle                4.0.1                      py_0
pyflakes                  2.1.1                    py37_0
pygments                  2.4.2                      py_0
pylint                    2.4.4                    py37_0
pympler                   0.7                        py_0
pynacl                    1.3.0            py37h62dcd97_0
pyopenssl                 19.1.0                   py37_0
pyparsing                 2.4.5                      py_0
pyqt                      5.9.2            py37h6538335_2
pyrsistent                0.15.6           py37he774522_0
pysocks                   1.7.1                    py37_0
python                    3.7.5                h8c8aaf0_0
python-dateutil           2.8.1                      py_0
python-jsonrpc-server     0.3.2                      py_0
python-language-server    0.31.2                   py37_0
pytz                      2019.3                     py_0
pywin32                   223              py37hfa6e2cd_1
pywin32-ctypes            0.2.0                    py37_0
pyxdg                     0.26                       py_0    spyder-ide
pyyaml                    5.1.2            py37he774522_0
pyzmq                     18.1.0           py37ha925a31_0
qdarkstyle                2.7                        py_0    spyder-ide
qt                        5.9.7            vc14h73c81de_0
qtawesome                 0.6.0                      py_0
qtconsole                 4.6.0                      py_0
qtpy                      1.9.0                      py_0
requests                  2.22.0                   py37_0
rope                      0.14.0                     py_0
rtree                     0.8.3                    py37_0
setuptools                42.0.1                   py37_0
sip                       4.19.8           py37h6538335_0
six                       1.13.0                   py37_0
snowballstemmer           2.0.0                      py_0
sortedcontainers          2.1.0                    py37_0
sphinx                    2.2.1                      py_0
sphinxcontrib-applehelp   1.0.1                      py_0
sphinxcontrib-devhelp     1.0.1                      py_0
sphinxcontrib-htmlhelp    1.0.2                      py_0
sphinxcontrib-jsmath      1.0.1                      py_0
sphinxcontrib-qthelp      1.0.2                      py_0
sphinxcontrib-serializinghtml 1.1.3                      py_0
spyder                    4.0.1                    py37_0
spyder-kernels            1.8.1                    py37_0
sqlite                    3.30.1               he774522_0
testpath                  0.4.4                      py_0
tornado                   6.0.3            py37he774522_0
traitlets                 4.3.3                    py37_0
ujson                     1.35             py37hfa6e2cd_0
urllib3                   1.24.2                   py37_0
vc                        14.1                 h0510ff6_4
vs2015_runtime            14.16.27012          hf0eaf9b_0
watchdog                  0.9.0                    py37_1    spyder-ide
wcwidth                   0.1.7                    py37_0
webencodings              0.5.1                    py37_1
wheel                     0.33.6                   py37_0
win_inet_pton             1.1.0                    py37_0
wincertstore              0.2                      py37_0
wrapt                     1.11.2           py37he774522_0
yaml                      0.1.7                hc54c509_2
yapf                      0.28.0                     py_0
zeromq                    4.3.1                h33f27b4_3
zipp                      0.6.0                      py_0
zlib                      1.2.11               h62dcd97_3

```

</details>

I could reproduce it in a brand new test env with just spyder-kernels, i.e. ``conda create -n test-env spyder-kernels``

<details>

```
    backcall:         0.1.0-py37_0
    ca-certificates:  2019.11.27-0
    certifi:          2019.11.28-py37_0
    cloudpickle:      1.2.2-py_0
    colorama:         0.4.3-py_0
    decorator:        4.4.1-py_0
    ipykernel:        5.1.3-py37h39e3cac_0
    ipython:          7.11.1-py37h39e3cac_0
    ipython_genutils: 0.2.0-py37_0
    jedi:             0.15.2-py37_0
    jupyter_client:   5.3.4-py37_0
    jupyter_core:     4.6.1-py37_0
    libsodium:        1.0.16-h9d3ae62_0
    openssl:          1.1.1d-he774522_3
    parso:            0.5.2-py_0
    pickleshare:      0.7.5-py37_0
    pip:              19.3.1-py37_0
    prompt_toolkit:   3.0.2-py_0
    pygments:         2.5.2-py_0
    python:           3.7.6-h60c2a47_1
    python-dateutil:  2.8.1-py_0
    pywin32:          227-py37he774522_0
    pyzmq:            18.1.0-py37ha925a31_0
    setuptools:       44.0.0-py37_0
    six:              1.13.0-py37_0
    spyder-kernels:   1.8.1-py37_0
    sqlite:           3.30.1-he774522_0
    tornado:          6.0.3-py37he774522_0
    traitlets:        4.3.3-py37_0
    vc:               14.1-h0510ff6_4
    vs2015_runtime:   14.16.27012-hf0eaf9b_1
    wcwidth:          0.1.7-py37_0
    wheel:            0.33.6-py37_0
    wincertstore:     0.2-py37_0
    zeromq:           4.3.1-h33f27b4_3
```

</details>

ccordoba12 commented:
> File "C:\Miniconda3\envs\spyder𔂮\lib\site‑packages\jupyter_client\launcher.py", line 138, in launch_kernel
> proc = Popen(cmd, **kwargs)
> File "C:\Miniconda3\envs\spyder𔂮\lib\subprocess.py", line 800, in __init__
restore_signals, start_new_session)
> PermissionError: [WinError 5] Access is denied

I think this problem is related to Spyder being run from @CAM-Gerlach's home directory, but jupyter_client being installed in a directory for which @CAM-Gerlach's regular user don't have permissions (i.e. in `C:\Miniconda3`).

@goanpeca, please take a look at jupyter_client to see if it's possible to disable that feature, so we can make that the case when running Spyder from `bootstrap.py`.

ccordoba12 commented on spyder/plugins/ipythonconsole/scripts/spyder-kernels.bat:
```diff
@@ -0,0 +1,13 @@
+:: This scripts helps activate a conda environment before running a spyder-kernel
```
Please rename this script to `conda-activate.bat`

goanpeca replied:
Done

ccordoba12 commented on spyder/plugins/ipythonconsole/scripts/spyder-kernels.bat:
```diff
@@ -0,0 +1,13 @@
+:: This scripts helps activate a conda environment before running a spyder-kernel
+@echo off
+
+set CODEPAGE=$1
```
Why is this `$1` instead of `%1`?

goanpeca replied:
Whoops

goanpeca replied:
Oops!

ccordoba12 commented on spyder/plugins/ipythonconsole/scripts/spyder-kernels.sh:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
Please rename this script to `conda-activate.sh`

goanpeca replied:
This name might suggest this is more geberal so it should be more like conda-activate-kernel or something

ccordoba12 commented on spyder/plugins/ipythonconsole/utils/kernelspec.py:
```diff
@@ -62,13 +94,43 @@ def argv(self):
+            # activation script to correctly activate the spyder-kernel
+
+            # If changes are needed on this section make sure you also update
+            # the activation scripts at spyder.plugins.ipythonconsole.scripts
```
Please change the dots here for slashes:

    spyder/plugins/ipythonconsole/scripts

ccordoba12 commented on spyder/utils/conda.py:
```diff
@@ -0,0 +1,76 @@
+    return '{quotes}{path}{quotes}'.format(quotes=quotes, path=path)
+
+
+def is_conda():
```
A couple of comments about this function:

1. Please name it `is_conda_env`.
2. This function needs to receive the path of the Python interpreter in the env (e.g. `/home/carlos/miniconda3/envs/py37/bin/python`), select the directory two levels above it (i.e. `/home/carlos/miniconda3/envs/py37`), which corresponds to `sys.prefix` in that env, and use that directory instead of `sys.prefix` below. That's because right now  you're detecting if Spyder is installed in a conda env, not that the interpreter we're using to start kernels is.

goanpeca replied:
Fixed!

ccordoba12 commented on spyder/utils/conda.py:
```diff
@@ -0,0 +1,76 @@
+    """
+    env_key = '{0}envs{0}'.format(os.sep)
+    if sys.prefix.rfind(env_key) != -1:
+        root_prefix = sys.prefix.split(env_key)[0]
```
I think this wouldn't work if Spyder is installed in a different environment from the base env.

goanpeca replied:
That is why there is an else :-)

goanpeca replied:
if `/envs/` is not found (== -1) use sys.prefix otherwise it is a non base env.

CAM-Gerlach commented:
@ccordoba12 That was my initial assumption, but I installed it as a regular user under my user account and verified the Miniconda3 root directory and all its subdirs have Full Control (rwx) permissions granted recursively to all logged in users, and as mentioned I tried launching Spyder as admin under my same user account from an admin command prompt, and it didn't make any difference. As far as I'm aware I haven't had similar permissions issues with anything else over many years of use.

I would install another copy to my user dir to make 100% sure but I'm on a really slow internet connection right now so it would take forever, and my user dir has spaces which could cause its own issues. I can try that when I get home tonight if you haven't figured it out by then.

pep8speaks commented:
Hello @goanpeca! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`spyder/utils/conda.py`](https://github.com/spyder-ide/spyder/blob/9fa621c8b59c73ecc1f18816ecc8c29ca07c1229/spyder/utils/conda.py):

> [Line 23:10](https://github.com/spyder-ide/spyder/blob/9fa621c8b59c73ecc1f18816ecc8c29ca07c1229/spyder/utils/conda.py#L23): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 23:10](https://github.com/spyder-ide/spyder/blob/9fa621c8b59c73ecc1f18816ecc8c29ca07c1229/spyder/utils/conda.py#L23): [E117](https://duckduckgo.com/?q=pep8%20E117) over-indented
> [Line 29:1](https://github.com/spyder-ide/spyder/blob/9fa621c8b59c73ecc1f18816ecc8c29ca07c1229/spyder/utils/conda.py#L29): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

goanpeca commented:
Thanks for the initial tests @CAM-Gerlach. I can reproduce the problem and will dig a bit into what is happening.

CAM-Gerlach commented:
@goanpeca Sure thing, thanks for working on this Gonzalo. I'm sure its some crazy stupid Windows quirk. I'm at the airport right now but let me know when you want me to test it again and I'll do my best to turn it around within a few hours.

goanpeca commented:
@CAM-Gerlach made some changes and now it seems to be working on my side. Could you test again :-) ?

goanpeca commented:
Added the tests @ccordoba12. @CAM-Gerlach let us know if this fixes the problems :-)

ccordoba12 commented:
@goanpeca, probably some slow tests on Windows fail because it takes too much time to start a kernel now. So you should increase the timeout of `qtbot.waitUntil` for those tests.

CAM-Gerlach commented:
@goanpeca Finally testing now, really sorry for the delay...

ccordoba12 commented:
These are great news @CAM-Gerlach!! And sorry for the big delay in fixing this one (one year and a half), but it was very tricky to do it correctly. Thankfully we could count with @goanpeca's expertise maitaining Anaconda Navigator to help us solve it.

ccordoba12 commented on spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py:
```diff
@@ -71,6 +72,22 @@ def get_console_background_color(style_sheet):
     return background_color
 
 
+def get_conda_test_env(test_env_name='jedi-test-env'):
```
Please add a docstring to describe what this helper function does.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py:
```diff
@@ -1567,5 +1596,24 @@ def test_calltip(ipyconsole, qtbot):
+@flaky(max_runs=3)
+@pytest.mark.test_environment_interpreter
+def test_conda_env_activation(ipyconsole, qtbot):
+    """Test that the conda environment of the console is activated."""
```
This should be

    Test that the conda environment associated with an external interpreter
    is activated before a kernel is created for it.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/plugins/ipythonconsole/utils/kernelspec.py:
```diff
@@ -62,13 +94,41 @@ def argv(self):
+                    get_activation_script(),  # This is bundled with Spyder
+                    codepage,
+                    get_conda_activation_script(pyexec),  # Might be remote
+                    get_conda_env_path(pyexec),  # Might be remote
```
remote -> external

Remote is a different thing.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/plugins/ipythonconsole/utils/kernelspec.py:
```diff
@@ -62,13 +94,41 @@ def argv(self):
+                kernel_cmd = [
+                    get_activation_script(),  # This is bundled with Spyder
+                    get_conda_activation_script(pyexec),  # Might be remote
+                    get_conda_env_path(pyexec),  # Might be remote
```
remote -> external

goanpeca replied:
Fixed

ccordoba12 commented on spyder/plugins/ipythonconsole/utils/manager.py:
```diff
@@ -0,0 +1,87 @@
+import psutil
+
+
+class SpyderKernelManager(QtKernelManager):
```
Please add a docstring for this class and add below comments above the lines you added to make this work with our activation scripts.

goanpeca replied:
Fixed

ccordoba12 commented on spyder/utils/conda.py:
```diff
@@ -0,0 +1,89 @@
+    If `quote` is True, then quotes are added if spaces are found in the path.
+    """
+    if os.name == 'nt':
+        activate = 'Scripts\\activate'
```
Shouldn't be safer to use here

    activate = 'Scripts' + osp.sep + 'activate'

?

goanpeca replied:
Fixed

ccordoba12 commented on spyder/utils/conda.py:
```diff
@@ -0,0 +1,89 @@
+    if os.name == 'nt':
+        activate = 'Scripts\\activate'
+    else:
+        activate = 'bin/activate'
```
Same here:

    activate = 'bin' + osp.sep + 'activate'

goanpeca replied:
Fixed

goanpeca commented:
>last comments, then this should be ready!

Fixed!

CAM-Gerlach commented:
@goanpeca @ccordoba12 All of the branches for the platform-specific path seperators are unnecessary, since the windows command line and API handles `/` and `\` equally, so the best practice is simply to always use `/` to minimize code complexity, maximize cross-platform compatibility and avoid any unintended edge cases with the wrong branch being taken.

ccordoba12 commented:
> All of the branches for the platform-specific path seperators are unnecessary, since the windows command line and API handles / and \ equally, so the best practice is simply to always use / to minimize code complexity, maximize cross-platform compatibility and avoid any unintended edge cases with the wrong branch being taken.

Agreed. @goanpeca, please address this suggestion.

ccordoba12 commented:
@goanpeca, I also have a question about this: how things work when a conda env path has unicode characters on it?

goanpeca commented:
Oops, I &%*^*&% the rebase. Will fix!

goanpeca commented:
I don't understand what happened @ccordoba12  I rebased against 4.x ... did you pull some master change into 4.x ?

goanpeca commented:
Ready for one more review round @ccordoba12 

Added tests for activation of environment with unicode chars. Changed the way we tests that an executable is a python executable.
